File: java/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return ans;
        }
        Arrays.sort(candidates);
        dfs(candidates, 0, target, new ArrayList<>(), ans);
        return ans;
    }
    private void dfs(int[] nums, int start, int target, List<Integer> temp, List<List<Integer>> ans) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        for (int i = start; i < nums.length; i++) {
            temp.add(nums[i]);
            dfs(nums, i, target - nums[i], temp, ans);
            temp.remove(temp.size() - 1);
        }
    }
}

File: python3/39.combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        
        def dfs(index, rest, temp):
            if rest == 0:
                ans.append(translate(temp, candidates))
                return
            if index == len(candidates) or rest < 0:
                return
            
            for time in range(rest // candidates[index] + 1):
                temp.append(time)
                dfs(index + 1, rest - time * candidates[index], temp)
                temp.pop()
                
        dfs(0, target, [])
        return ans
    
def translate(times, candidates):
    ans = []
    for i in range(len(times)):
        ans.extend([candidates[i]] * times[i])
    return ans
                
        
